book/accessibility.md

```diff
@@ -2186,15 +2186,45 @@ bounds:
 class AccessibilityNode:
     def __init__(self, node):
         # ...
-        if node.layout_object:
-            self.bounds = absolute_bounds_for_obj(node.layout_object)
-        else:
-            self.bounds = None
+        self.bounds = self.compute_bounds()
+
+    def compute_bounds(self):
+        if self.node.layout_object:
+            return [absolute_bounds_for_obj(self.node.layout_object)]
+        # ...
 ```
 
-Note that I'm using `absolute_bounds_for_obj` here, because the bounds
-we're interested in are the absolute coordinates on the screen, after
-any transformations like `translate`.
+Note that I'm using `absolute_bounds_for_obj` here, because the bounds we're
+interested in are the absolute coordinates on the screen, after any
+transformations like `translate`.
+
+However, there is another complication: it may not be that `node.layout_object`
+is set; for example, text nodes do not have one.^[And that's ok, because I
+chose not to set bounds at all for these nodes, as they are not focusable.]
+Likewise, nodes with inline layout generally do not. So we need to walk up the
+tree to find the parent with a `BlockLayout` and union all text nodes in all
+`LineLayouts` that are children of the current `node`. And because there can be
+multiple `LineLayouts` and text nodes, the bounds needs to be an array of
+`skia.Rect` objects:
+
+``` {.python}
+class AccessibilityNode:
+    def compute_bounds(self):
+        # ...
+        if isinstance(self.node, Text):
+            return []
+        inline = self.node.parent
+        bounds = []
+        while not inline.layout_object: inline = inline.parent
+        for line in inline.layout_object.children:
+            line_bounds = skia.Rect.MakeEmpty()
+            for child in line.children:
+                if child.node.parent == self.node:
```

Comment:
So this is disappointing, because it's going to misbehave the same way as outlines on lines in general, but ultimately I guess this is fine, it's an issue we already have.